gtsfm/densify/mvs_patchmatchnet.py

```diff
@@ -0,0 +1,247 @@
+        else:
+            state_dict = torch.load(PATCHMATCHNET_WEIGHTS_PATH, map_location=torch.device("cpu"))
+
+        model.load_state_dict(state_dict["model"])
+        model.eval()
+
+        depth_est_list = {}
+        confidence_est_list = {}
+        with torch.no_grad():
+            for batch_idx, sample in enumerate(loader):
+                start_time = time.time()
+
+                # check if cuda devices is supported, and store the inference data to the target device
+                if torch.cuda.is_available():
+                    sample_device = tocuda(sample)
+                else:
+                    sample_device = copy.copy(sample)
+
+                # inference using PatchmatchNet
+                outputs = model(
+                    sample_device["imgs"],
+                    sample_device["proj_matrices"],
+                    sample_device["depth_min"],
+                    sample_device["depth_max"],
+                )
+
+                outputs = tensor2numpy(outputs)
+                del sample_device
+
+                ids = sample["idx"]
+
+                # save depth maps and confidence maps
+                for idx, depth_est, photometric_confidence in zip(
+                    ids, outputs["refined_depth"]["stage_0"], outputs["photometric_confidence"]
+                ):
+
+                    idx = idx.cpu().numpy().tolist()
+                    depth_est_list[idx] = depth_est.copy()
+                    confidence_est_list[idx] = photometric_confidence.copy()
+
+                logger.info(
+                    "[Densify::PatchMatchNet] Iter {}/{}, time = {:.3f}".format(
+                        batch_idx + 1, len(loader), time.time() - start_time
+                    )
+                )
+
+        # filter inference result with thresholds
+        dense_point_cloud = self.filter_depth(
+            dataset=dataset,
+            depth_list=depth_est_list,
+            confidence_list=confidence_est_list,
+            geo_pixel_thres=thresholds[0],
+            geo_depth_thres=thresholds[1],
+            photo_thres=thresholds[2],
+        )
+
+        return dense_point_cloud
+
+    def filter_depth(
+        self,
+        dataset: PatchmatchNetData,
+        depth_list: Dict[int, np.ndarray],
+        confidence_list: Dict[int, np.ndarray],
+        geo_pixel_thres: float,
+        geo_depth_thres: float,
+        photo_thres: float,
+    ) -> np.ndarray:
+        """Filter depth map and get filtered dense point cloud
+        Ref: Wang et al. https://github.com/FangjinhuaWang/PatchmatchNet/blob/main/eval.py
+
+        Args:
+            dataset (PatchmatchNetData): an instance of PatchmatchData as the inference dataset
+            depth_list (Dict[int, np.ndarray]): list of 2D depth map (H, W) from each view
+            confidence_list (Dict[int, np.ndarray]): list of 2D confidence map (H, W) from each view
+            geo_pixel_thres (float): geometric pixel threshold
```

Comment:
no need to repeat the types down here, since they're already clear from the function signature

Reply:
@codyly could you take another look at my comments here?